### PR TITLE
added methods to identify server errors

### LINKED ISSUE #35: Non-typed Errors Returned
Currently, errors from the LDAP server are returned untyped. It would be useful to have a separate type for each error. I'll work on this.

### PR DESCRIPTION
**Pull request by stevekuznetsov**
I've added exposed methods that can identify server errors being given from this package. I've also made more use of the `getLDAPResultCode` method, as in some places the extraction was still being done without the method. I'm wondering if some sort of reflection-based unit test would be useful for these methods.

Fixes #35 

@liggitt @johnweldon PTAL

### CODE REVIEW

File: error.go [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+		response := packet.Children[1]
+		if response.ClassType == ber.ClassApplication && response.TagType == ber.TypeConstructed && len(response.Children) >= 3 {
+			return LDAPServerResultCode(response.Children[0].Value.(int64)), response.Children[2].Value.(string)
+		}
+	}
+
+	return ErrorNetwork, "Invalid packet format"
+}
+
+type LDAPServerError struct {
```
Comment (liggitt) [inline error.go]: this looks exactly like the other error struct... why is this needed?

Comment (liggitt) [inline error.go]: actually, I like the other struct better... the internal error and result code are directly accessible

Comment (liggitt) [inline error.go]: also, this will make anyone using this package rework their code for no reason

File: error.go [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+
+func (e *LDAPServerError) Error() string {
+	return fmt.Sprintf("LDAP Result Code %d %q: %s", e.resultCode, LDAPResultCodeMap[e.resultCode], e.err.Error())
+}
+
+func NewServerError(resultCode LDAPServerResultCode, err error) error {
+	return &LDAPServerError{resultCode: resultCode, err: err}
+}
+
+func isSpecificLDAPServerError(err error, desiredResultCode LDAPServerResultCode) bool {
```
Comment (liggitt) [inline error.go]: I'd actually rather use this helper than a thousand unique helper methods... can we export this one instead?

File: error.go (hunk 1 above)
Comment (stevekuznetsov) [inline error.go]: I'll revert this.

File: error.go (hunk 2 above)
Comment (stevekuznetsov) [inline error.go]: I like not thinking about return codes, but I can see how it's little different than this.

**Comment by stevekuznetsov** [comment]
@liggitt Addressed comments

File: conn.go [hunk 3]
```diff
@@ -177,7 +177,7 @@ func (l *Conn) StartTLS(config *tls.Config) error {
 		ber.PrintPacket(packet)
 	}
 
-	if packet.Children[1].Children[0].Value.(int64) == LDAPResultSuccess {
```
Comment (liggitt) [inline conn.go]: don't look this up twice, do
```
if resultCode, message := getLDAPResultCode(packet); resultCode == LDAPResultSuccess {
  ...
} else {
  return NewError(resultCode, fmt.Errorf("ldap: cannot StartTLS (%s)", message))
}
```

File: conn.go [hunk 4]
```diff
@@ -190,9 +190,8 @@ func (l *Conn) StartTLS(config *tls.Config) error {
 	} else {
 		// https://tools.ietf.org/html/rfc4511#section-4.1.9
 		// Children[1].Children[2] is the diagnosticMessage which is guaranteed to exist.
```
Comment (liggitt) [inline conn.go]: move this comment into getLDAPResultCode

File: error.go [hunk 5]
```diff
@@ -0,0 +1,142 @@
+package ldap
+
+import (
+	"fmt"
+
+	"gopkg.in/asn1-ber.v1"
+)
+
+type LDAPServerResultCode uint8
```
Comment (liggitt) [inline error.go]: Don't change the type, that can break clients using this library. I like the typed constants, but it's not worth a breaking change now.

File: error.go [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+func getLDAPResultCode(packet *ber.Packet) (code LDAPServerResultCode, description string) {
+	if len(packet.Children) >= 2 {
+		response := packet.Children[1]
+		if response.ClassType == ber.ClassApplication && response.TagType == ber.TypeConstructed && len(response.Children) >= 3 {
+			return LDAPServerResultCode(response.Children[0].Value.(int64)), response.Children[2].Value.(string)
+		}
+	}
+
+	return ErrorNetwork, "Invalid packet format"
```
Comment (liggitt) [inline error.go]: would `ErrorUnexpectedResponse` be more appropriate?

Comment (liggitt) [inline error.go]: nevermind, this is a straight move

File: error.go [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+	if err == nil {
+		return false
+	}
+
+	serverError, ok := err.(*Error)
+	if !ok {
+		return false
+	}
+
+	if serverError.ResultCode == desiredResultCode {
```
Comment (liggitt) [inline error.go]: `return serverError.ResultCode == desiredResultCode`

Comment (stevekuznetsov) [inline error.go]: Embarrassing

**Comment by stevekuznetsov** [comment]
@liggitt Addressed comments

File: error.go [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+	LDAPResultAffectsMultipleDSAs:          "Affects Multiple DSAs",
+	LDAPResultOther:                        "Other",
+}
+
+func getLDAPResultCode(packet *ber.Packet) (code uint8, description string) {
+	if len(packet.Children) >= 2 {
+		response := packet.Children[1]
+		if response.ClassType == ber.ClassApplication && response.TagType == ber.TypeConstructed && len(response.Children) >= 3 {
+			// Children[1].Children[2] is the diagnosticMessage which is guaranteed to exist as seen here: https://tools.ietf.org/html/rfc4511#section-4.1.9
+			return response.Children[0].Value.(uint8), response.Children[2].Value.(string)
```
Comment (liggitt) [inline error.go]: The serialized value is a uint64. You have to do `uint8(response.Children[0].Value.(uint64))`. Tests will catch this, make sure you have a clean run

Comment (stevekuznetsov) [inline error.go]: `go test` came up clean with this implementation. I'll update the cast.

**Comment by stevekuznetsov** [comment]
@liggitt Ready

**Comment by liggitt** [comment]
LGTM, thanks
